Add files_external:list command to list configured external storages

icewind1991 opened the pull request:
First part of #20368

Usage:

```
occ files_external:list [--json] [user_id]
```

Example output:

```
+----------+-------------+---------+-----------------------+--------------------------------------------------------------------------------------------------------------+----------------+------------------+-------------------+
| Mount ID | Mount Point | Storage | Authentication Type   | Configuration                                                                                                | Options        | Applicable Users | Applicable Groups |
+----------+-------------+---------+-----------------------+--------------------------------------------------------------------------------------------------------------+----------------+------------------+-------------------+
| 1        | /Pictures   | Local   | None                  | datadir: "\/home\/robin\/Pictures"                                                                           | encrypt: false | icewind          |                   |
| 2        | /dav        | WebDAV  | Username and password | host: "localcloud\/owncloud\/remote.php\/webdav", root: "doc", secure: false, user: "test", password: "test" |                | icewind, test    |                   |
+----------+-------------+---------+-----------------------+--------------------------------------------------------------------------------------------------------------+----------------+------------------+-------------------+

```

cc @PVince81 @DeepDiver1975 @Xenopathic

icewind1991 committed:
Add files_external:list command to list configured external storages

PVince81 commented on apps/files_external/appinfo/register_command.php:
```diff
@@ -0,0 +1,34 @@
+<?php
+/**
+ * @author Björn Schießle <schiessle@owncloud.com>
```
Oh hello Björn :smile:

PVince81 commented:
Nice.

- [ ] Can you obfuscate / remove the password from the displayed list ?

Other wise it looks good.

Xenopathic commented on apps/files_external/command/listcommand.php:
```diff
@@ -0,0 +1,193 @@
+				InputArgument::OPTIONAL,
+				'user id to list the personal mounts for, if no user is provided admin mounts will be listed'
+			)
+			->addOption('json', null, InputOption::VALUE_NONE, 'use json output instead of a human-readable array');
```
Symfony Console already supports the `--output=...` option, we should use that instead of rolling our own parameter.

icewind1991 replied:
done

icewind1991 committed:
use common --output option

icewind1991 commented:
> Can you obfuscate / remove the password from the displayed list ?

I'm not sure if we want that, imo for the admin it's useful to see the passwords you configured

Xenopathic commented on apps/files_external/command/listcommand.php:
```diff
@@ -0,0 +1,197 @@
+
+	protected function configure() {
+		$this
+			->setName('files_external:list')
```
Can we use `files-external:list` instead? All other commands use dashes as word separators, not underscores

icewind1991 replied:
I went for underscores since that's the app id

Xenopathic commented on apps/files_external/command/listcommand.php:
```diff
@@ -0,0 +1,197 @@
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class ListCommand extends Base {
```
For consistency with how other commands are defined, this should be `class List` (assuming that doesn't conflict with a language keyword?)

icewind1991 replied:
> assuming that doesn't conflict with a language keyword

It does

Xenopathic commented:
In it's current form (without output line limiting), it's not the most useful command:

```
+----------+-------------+---------+---------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+
| Mount ID | Mount Point | Storage | Authentication Type | Configuration                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       | Options |
+----------+-------------+---------+---------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+
| 2        | /SFTP       | SFTP    | RSA public key      | host: "nyx.lan.mccorkell.me.uk", root: "\/home\/test", user: "test", public_key: "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAAAgQCp0mg6Tg1V+WmkaHsgViwsQBlwETF6QVKR0FqNM5n3kXu+LGvdakvHfNK0B+pPukTnMZ\/oN6+kEch5OQD5\/K0EprM5PhWp7I1g4uTgzsDyJQmYxgFdGE+\/CY2nhUiz\/yBEnJoQFJUyHL9JsRX8UdGstA57JeLchdykL4em0ZtAnw== phpseclib-generated-key", private_key: "-----BEGIN RSA PRIVATE KEY-----\nProc-Type: 4,ENCRYPTED\nDEK-Info: DES-EDE3-CBC,024FAF6682942FF5\n\n6wbWKjZZNSE2\/X2Ai5ldrLvAy6n1r64RMBcsZOPyxm0ut9x1lzj8SaNETo2Jv\nvwaOJcGrV57a39NZGK4qJx6hfH+wH0LEdLJEGwunOwNxS32xr2NURqkFB0ZNSx\nAnQ0qO0zswjokKwfTmzGXIE7KuMC+A\/JXXaCHoTrbuXGQ5Gr3dqhp9aoqPrvHfil\n4859jfrapcyTDe\/V5kugFHeRGWZ68o2mhXJM4z4b9QNDifKbO7HoHWT4W6Lzz\nbNZ0vtIeeabka8a1stoptryingtohackme0Jfs6onIHsDK4FjbAmp2dA7ruh\/NZVvfVFoFanm1ahB\nqklMQFJWzlF7l+A9jHwLe7k\/MZZVmYFh4G5iAWPUUH2ccf7dm+tSsybncI3Ly+\nz0EXg96DiyZPDdiqHU4Q8ktrUlzsbCfQJOxoBjKkSnA\/mMpIJQXf3J2JvoKU1\nHwp5rtOltcbNJO44YoNsVcSlWQGHZ06M9hz1+KD7t8shX4SkAub3WMR6mjpprC\nM1S89ZP9Cmpq9LrFET7F+3daEQY3dPaXZro\/HFpSd1UeizfRbrOQJ\nX0tf3IrApW69M2POUznk55qs+LE19tT+BdxHDjjiR0cb\/i3sMJ2oECOmiDcKExd\noSPX0NtiyKNusEtdlsUIS1DCxcqm4Cc+bc7h6qDoosWNa7TY4oUbErRl4dhx0Lxg\nlTzvq8t3qHqIwioLyB+tgfO61SOa7TY7lyhS\/hpsuxMDOUoPAmzXbrnN4zVT8\/Ck\ncGX+r\/5V1RX4BzNdPcSO7K8SXMBaIHZaPo6iB4+0Z\/CQs3tkSBV48Q==\n-----END RSA PRIVATE KEY-----" |         |
+----------+-------------+---------+---------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+
```

Perhaps see how the `ldap:list-config` command does it?

icewind1991 committed:
hide passwords and secrets on default

icewind1991 committed:
truncate long values on default

icewind1991 commented:
Password and other secrets are not hidden by default and long values are truncated

- Added `--show-password` option to show passwords and secrets
- Added `--full` op option to show long config values

icewind1991 commented:
Example of password hiding and truncating

```
+----------+-------------+--------------+---------------------+--------------------------------------------------------------------------------------+----------------+------------------+-------------------+
| Mount ID | Mount Point | Storage      | Authentication Type | Configuration                                                                        | Options        | Applicable Users | Applicable Groups |
+----------+-------------+--------------+---------------------+--------------------------------------------------------------------------------------+----------------+------------------+-------------------+
| 1        | /asdasdasd  | Google Drive | OAuth2              | configured: "true", client_id: "403771...nt.com", client_secret: "***", token: "***" |                | All              |                   |
+----------+-------------+--------------+---------------------+--------------------------------------------------------------------------------------+----------------+------------------+-------------------+
```

Xenopathic commented:
Very nice! Some language fixes incoming though...

Xenopathic commented on apps/files_external/command/listcommand.php:
```diff
@@ -0,0 +1,231 @@
+				'full',
+				null,
+				InputOption::VALUE_NONE,
+				'dont truncate long values in table output'
```
don't

Xenopathic commented on apps/files_external/command/listcommand.php:
```diff
@@ -0,0 +1,231 @@
+			if ($userId) {
+				$output->writeln("<info>No mounts configured by $userId</info>");
+			} else {
+				$output->writeln("<info>No mounts admin configured</info>");
```
No admin mounts configured

icewind1991 committed:
language fixes

icewind1991 commented:
>language  fixes incoming though...

Fixed

Xenopathic commented:
:+1:

PVince81 commented:
@icewind1991 it doesn't show the mounts of individual users, something for later ?

:+1:

PVince81 commented:
also the JSON doesn't look the same as the old mount.json. It's fine like this as long as the later import function supports both formats.

icewind1991 commented:
> it doesn't show the mounts of individual users, something for later ?

Specify the user as argument

```
occ files_external:list icewind
```
